numpyro/infer/enum_messenger.py

```diff
@@ -0,0 +1,577 @@
+
+    def __enter__(self):
+        super().__enter__()  # do this first to take care of globals recycling
+        name_to_dim = OrderedDict([(self.name, self.dim)]) if self.dim is not None else OrderedDict()
+        indices = to_data(self._indices, name_to_dim=name_to_dim, dim_type=DimType.VISIBLE)
+        # extract the dimension allocated by to_data to match plate's current behavior
+        self.dim, self.indices = -indices.dim(), indices.squeeze()
+        return self
+
+    @staticmethod
+    def _get_batch_shape(cond_indep_stack):
+        n_dims = max(-f.dim for f in cond_indep_stack)
+        batch_shape = [1] * n_dims
+        for f in cond_indep_stack:
+            batch_shape[f.dim] = f.size
+        return tuple(batch_shape)
+
+    def process_message(self, msg):  # copied almost verbatim from plate
+        if msg['type'] not in ('sample',):
+            return
+        cond_indep_stack = msg['cond_indep_stack']
+        frame = CondIndepStackFrame(self.name, self.dim, self.subsample_size)
+        cond_indep_stack.append(frame)
+        expected_shape = self._get_batch_shape(cond_indep_stack)
+        dist_batch_shape = msg['fn'].batch_shape if msg['type'] == 'sample' else ()
+        overlap_idx = max(len(expected_shape) - len(dist_batch_shape), 0)
+        trailing_shape = expected_shape[overlap_idx:]
+        # e.g. distribution with batch shape (1, 5) cannot be broadcast to (5, 5)
+        broadcast_shape = lax.broadcast_shapes(trailing_shape, dist_batch_shape)
+        if broadcast_shape != dist_batch_shape:
+            raise ValueError('Distribution batch shape = {} cannot be broadcast up to {}. '
+                             'Consider using unbatched distributions.'
+                             .format(dist_batch_shape, broadcast_shape))
+        batch_shape = expected_shape[:overlap_idx]
+        if 'sample_shape' in msg['kwargs']:
+            batch_shape = lax.broadcast_shapes(msg['kwargs']['sample_shape'], batch_shape)
+        msg['kwargs']['sample_shape'] = batch_shape
+        if self.size != self.subsample_size:
+            scale = 1. if msg['scale'] is None else msg['scale']
+            msg['scale'] = scale * self.size / self.subsample_size
+
+
+class enum(BaseEnumMessenger):
+    """
+    This version of EnumMessenger uses to_data to allocate a fresh enumeration dim
+    for each discrete sample site.
+    """
+    def process_message(self, msg):
+
+        if msg["type"] != "sample" or \
+                msg["done"] or msg["is_observed"] or msg["infer"].get("expand", False) or \
+                msg["infer"].get("enumerate") != "parallel":
+            return
+
+        if msg["infer"].get("num_samples", None) is not None:
+            raise NotImplementedError("TODO implement multiple sampling")
+
+        if msg["infer"].get("expand", False):
+            raise NotImplementedError("expand=True not implemented")
+
+        raw_value = msg["fn"].enumerate_support(expand=False).squeeze()
```

Comment:
@fehiepsi does this functionality exist yet in NumPyro?  If not, can you suggest an alternative implementation for this line?

Reply:
Hmm..we didn't bother to add this since we didn't have enumeration, but it should be straightforward to add this to our discrete distributions. WDYT, @fehiepsi?

Reply:
I can just add a workaround here for now

Reply:
I'll add this soon.